snuba/subscriptions/scheduler_consumer.py

```diff
@@ -0,0 +1,313 @@
+    # correctly commit the offsets that represent the interval so that
+    # intervals are not repeated or skipped when a consumer restarts (or more
+    # likely rebalances, which can be generalized to a restart operation.)
+    #
+    # Take the previously described scenario where we had just returned a tick
+    # interval that was represented by the mesasges A and B: without taking any
+    # precautions, the Kafka consumer would use the next offset from message B
+    # for commit, which would be 2 in this case (1 + 1). If the consumer were
+    # to crash and restart, it would resume at offset 2, causing the next tick
+    # interval returned to be for the messages C and D -- in this case, B and C
+    # was never returned! To avoid skipping intervals when restarting, the
+    # consumer would have had to commit the offset 1 (the offset of message B)
+    # to ensure that upon restart, the next interval would be the interval
+    # between B and C, since the message B was the first message received by
+    # the consumer.
+    def __init__(
+        self, consumer: Consumer[KafkaPayload], time_shift: Optional[timedelta] = None
+    ) -> None:
+        self.__consumer = consumer
+        self.__previous_messages: MutableMapping[Partition, MessageDetails] = {}
+        self.__time_shift = time_shift if time_shift is not None else timedelta()
+
+    def subscribe(
+        self,
+        topics: Sequence[Topic],
+        on_assign: Optional[Callable[[Mapping[Partition, int]], None]] = None,
+        on_revoke: Optional[Callable[[Sequence[Partition]], None]] = None,
+    ) -> None:
+        self.__consumer.subscribe(topics, on_assign=on_assign, on_revoke=on_revoke)
+
+    def unsubscribe(self) -> None:
+        self.__consumer.unsubscribe()
+
+    def poll(self, timeout: Optional[float] = None) -> Optional[Message[Tick]]:
+        message = self.__consumer.poll(timeout)
+        if message is None:
+            return None
+
+        commit = commit_codec.decode(message.payload)
+        assert commit.orig_message_ts is not None
+
+        previous_message = self.__previous_messages.get(commit.partition)
+
+        result: Optional[Message[Tick]]
+        if previous_message is not None:
+            try:
+                time_interval = Interval(
+                    previous_message.orig_message_ts, commit.orig_message_ts
+                )
+            except InvalidRangeError:
+                logger.warning(
+                    "Could not construct valid time interval between %r and %r!",
+                    previous_message,
+                    MessageDetails(commit.offset, commit.orig_message_ts),
+                    exc_info=True,
+                )
+                return None
+            else:
+                result = Message(
+                    message.partition,
+                    message.offset,
+                    Tick(
+                        commit.partition.index,
```

Comment:
Yes its intentional since our message is dealing with 2 topics and their partition numbers now. The message is supposed to like a standard consumer (i.e. the messages' partition and offset is the one of the commit log). The tick contains the parts that are to do with the original message on `events`.